vm/src/gc.rs

```diff
@@ -540,14 +551,35 @@ impl Gc {
     }
 
     /// Allocates a new object.
-    pub fn alloc<D>(&mut self, def: D) -> GcPtr<D::Value>
+    pub fn alloc<D>(&mut self, def: D) -> Result<GcPtr<D::Value>>
+        where D: DataDef,
+              D::Value: Sized + Any
+    {
+        let size = def.size();
+        let needed = self.allocated_memory.saturating_add(size);
+        if needed >= self.memory_limit {
+            return Err(Error::OutOfMemory {
+                limit: self.memory_limit,
+                needed: needed,
+            });
+        }
+        Ok(self.alloc_ignore_limit_(size, def))
+    }
+
+    pub fn alloc_ignore_limit<D>(&mut self, def: D) -> GcPtr<D::Value>
```

Comment:
It is so error messages can be allocated even if the thread would otherwise be out of memory